src/stateDirectives.js

```diff
@@ -279,7 +279,61 @@ function $StateRefActiveDirective($state, $stateParams, $interpolate) {
   };
 }
 
+/**
+ * @ngdoc directive
+ * @name ui.router.state.directive:ui-state
+ *
+ * @requires ui.router.state.uiSref
+ *
+ * @restrict A
+ *
+ * @description
+ * Much like ui-sref, but will accept named $scope propertys to evaluate for a state definition,
+ * params and override options.
+ *
+ * @param {string} ui-state 'stateName' can be any valid absolute or relative state
+ * @param {Object} ui-state-params params to pass to {@link ui.router.state.$state#href $state.href()}
+ * @param {Object} ui-state-opts options to pass to {@link ui.router.state.$state#go $state.go()}
+ */
+$StateRefDynamicDirectve.$inject = [ 'uiSrefDirective' ];
+function $StateRefDynamicDirectve (uiSrefDirective) {
+  return {
+    require: uiSrefDirective[0].require,
+    link: function (scope, element, attrs) {
+      var args = arguments;
+
+      function buildUiSref () {
+        var baseRef = scope.$eval(attrs.uiState);
+
+        if (attrs.uiStateParams) {
+          baseRef += '(' + JSON.stringify(scope.$eval(attrs.uiStateParams)) + ')';
+        }
+
+        return baseRef;
+      }
+
+      function runStateRefLink () {
+        angular.extend(args[2], {
+          uiSref:     buildUiSref(),
+          uiSrefOpts: attrs.uiStateOpts,
+        });
+
+        uiSrefDirective[0].link.apply(null, args);
+      }
+
+      scope.$watch(function () {
+        return scope.$eval(attrs.uiState) +
```

Comment:
I like that, cutting away at the amount of $eval!

But - I do think we have to keep at least one $eval in there, otherwise changes wouldn't be picked up if the underlying data bound to the scope variable is modified - only so if the *actual* variable in the ui-state definition changes. 

So, my suggestion would be: 

```js
scope.$watch(function () {
  return scope.$eval([attrs.uiState, attrs.uiStateParams, attrs.uiStateOpts]);
}, runStateRefLink, true);
```

What do you think about that? @nateabele